### PR TITLE
Acquire read holds after the timestamp selection of a REFRESH MV

### PR DESCRIPTION
**Pull request by ggevay**
This fixes https://github.com/MaterializeInc/materialize/issues/24288 as discussed with @mjibson [here](https://materializeinc.slack.com/archives/C02FWJ94HME/p1704989974860729?thread_ts=1704982981.659269&cid=C02FWJ94HME).

### Motivation

  * This PR fixes a recognized bug: https://github.com/MaterializeInc/materialize/issues/24288

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: src/adapter/src/coord/read_policy.rs
Comment (mjibson) [inline src/adapter/src/coord/read_policy.rs:437]: The insert could delete some existing read holds. This should instead do something like `.entry(connid).or_insert_with(|| ReadHolds::new()).extend(read_holds)`

File: src/adapter/src/coord/command_handler.rs
Comment (mjibson) [inline src/adapter/src/coord/command_handler.rs:797]: Instead of a `bool` function argument, what about `acquire_read_holds_for: Option<&mut Session>`? That way the mut session is only passed if needed, and also serves as the flag to auto cleanup the read hold?

File: src/adapter/src/coord/command_handler.rs
Comment (mjibson) [inline src/adapter/src/coord/command_handler.rs:867]: Low preference thought, ignore if you like: if this has to be a vec, seems reasonable to require the caller pass a vec?

File: src/adapter/src/coord/command_handler.rs
Comment (mjibson) [inline src/adapter/src/coord/command_handler.rs:868]:
```suggestion
        let mut id_bundle = CollectionIdBundle::default();
```

File: src/adapter/src/coord/id_bundle.rs
Comment (mjibson) [inline src/adapter/src/coord/id_bundle.rs:29]: This can get removed, use default instead.

File: src/adapter/src/coord/command_handler.rs
Comment (mjibson) [inline src/adapter/src/coord/command_handler.rs:848]: Why is it correct to use all clusters here instead of the active cluster?

Comment (ggevay) [reply src/adapter/src/coord/command_handler.rs]: Well, I didn't want to move or duplicate the cluster selection logic here, which currently happens [later](https://github.com/MaterializeInc/materialize/blob/e4cce43ce8b4fe54c57fe8edf3db94d929952cad/src/sql/src/plan/statement/ddl.rs#L2140-L2143).

I think it shouldn't cause any problems if we have read holds on a few more objects on some other clusters for a short while. But if this is too scary from a workload isolation perspective (e.g., optimizing the MV might take forever, at least for now that we don't have a timeout on optimizer calls), then maybe I can move or duplicate the cluster selection logic here, and then just acquire head holds from the relevant cluster. What do you think?

File: src/adapter/src/coord/id_bundle.rs
Comment (ggevay) [reply src/adapter/src/coord/id_bundle.rs]: done

File: src/adapter/src/coord/command_handler.rs
Comment (ggevay) [reply src/adapter/src/coord/command_handler.rs]: Yes, thx!

File: src/adapter/src/coord/command_handler.rs
Comment (ggevay) [reply src/adapter/src/coord/command_handler.rs]: Ok, I don't really have a preference, so I've changed it.

File: src/adapter/src/coord/command_handler.rs
Comment (ggevay) [reply src/adapter/src/coord/command_handler.rs]: Good idea!

File: src/adapter/src/coord/read_policy.rs
Comment (ggevay) [reply src/adapter/src/coord/read_policy.rs]: Oh, yes, sorry!

Actually, I copied this line from [here](https://github.com/MaterializeInc/materialize/blob/d6a3c785559d40564f9b70fb8536a43ac1e8cdc0/src/adapter/src/coord/sequencer/inner/peek.rs#L694). So then this is probably a bug also on `main`?

I've fixed both occurrences now.

**Comment by ggevay** [comment]
There is some new, rather scary-looking error in the test that I added in the slt. I'll investigate tomorrow.

File: src/adapter/src/coord/command_handler.rs
Comment (mjibson) [reply src/adapter/src/coord/command_handler.rs]: I think it's worth doing that work to avoid possible isolation problems.

Comment (ggevay) [reply src/adapter/src/coord/command_handler.rs]: Ok, then I'll factor out that logic into a function, and call it also here.

**Comment by ggevay** [comment]
Actually, it turned out that I can trigger the above-mentioned error also on main. If I simply add
```
statement ok
select mz_unsafe.mz_sleep(3);
```
to the end of `materialized_views.slt` then I get this soft_panic:
```
frontier advancements after the `until` should be suppressed
```
I'll have to fix this now before merging this PR.

**Comment by ggevay** [comment]
(To unblock the slt run, I (possibly temporarily) pushed a commit that flips the `persist_txn_tables_cli` switch back in the slt runner config. Slack discussion [here](https://materializeinc.slack.com/archives/C01CFKM1QRF/p1705490015764119).)

File: src/adapter/src/coord/command_handler.rs
Comment (ggevay) [reply src/adapter/src/coord/command_handler.rs]: Unfortunately, this is turning out more painful than expected, because `resolve_cluster` is a method of `StatementContext`, and we don't yet have a `StatementContext` in the purification.

It uses the `StatementContext` just to access the catalog, so I can probably hack this somehow, but I'm not really sure whether it's worth it at this point. E.g., maybe I can make a `resolve_cluster` variation that takes a `&'a dyn SessionCatalog` instead of a `StatementContext`.

Comment (mjibson) [reply src/adapter/src/coord/command_handler.rs]: > maybe I can make a `resolve_cluster` variation that takes a `&'a dyn SessionCatalog`

Seems right to me. The StatementContext impl is just a wrapper around that anyway.

Comment (ggevay) [reply src/adapter/src/coord/command_handler.rs]: Ok, done!

**Comment by ggevay** [comment]
I've addressed all comments. Thanks for the review!

There is currently a commit that removes a `soft_panic`, I'll revert this after Dan's PR (https://github.com/MaterializeInc/materialize/pull/24492) is merged. (Dan's PR will solve the soft_panic for this PR, but we might still decide to remove the soft_panic later, depending of the outcome of [this discussion](https://materializeinc.slack.com/archives/CMNT2QSBZ/p1705509940786839). However, even if we ultimately decide to remove it, that should be done in a separate PR.)

File: src/adapter/src/coord/command_handler.rs
Comment (mjibson) [inline src/adapter/src/coord/command_handler.rs:837]: This can use `.catalog()` since it's executing in thread.

**Review by mjibson — APPROVED** [review]

File: src/compute/src/sink/refresh.rs
Comment (aalexandrov) [inline src/compute/src/sink/refresh.rs:118]: Why is this commented out?

Comment (ggevay) [reply src/compute/src/sink/refresh.rs:118]: See here: https://github.com/MaterializeInc/materialize/pull/24443#issuecomment-1896468096

Comment (ggevay) [reply src/compute/src/sink/refresh.rs]: I've remove this soft_panic in an other PR: https://github.com/MaterializeInc/materialize/pull/24582

**Pull request merged**
